### PR TITLE
Manual maintenance

### PR DESCRIPTION
**Pull request by SiegfriedSobkowiak**
# Manual Maintenance Sprint [NR]

Note: To create a PR using this template add the query parameter `template=manual_maintenance.md` to the merge request
creation URL (or simply copy this md file into the description)

# Maintenance tasks (to be completed by the assignee)

### Problematic updates

See the list of known issues on the internal Dataland Wiki.
Being present on this list does not mean that we simply skip the update, instead we are just aware that it may cause a problem.
If an issue arises from a new update that cannot be solved in the scope of MM, add it to the wiki page and create a ticket in the backlog.

### Gradle update

- [x] Execute `gradlew dependencyUpdates` to get a report on Dependencies with updates
- [x] Execute `refreshVersions` in Gradle tasks or `gradlew refreshVersions` to generate version suggestions in `versions.properties`
- [x] Update versions in `versions.properties`
- [x] Update the gradle wrapper: execute `gradlew wrapper --gradle-version X.Y.Z` with X.Y.Z being the latest released version

### Dataland frontend

- [x] Update node version in `dataland-frontend/build.gradle.kts`
- [x] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  `package.json`
- [x] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dataland keycloak theme

- [x] Update node version in `dataland-keycloak/dataland_theme/login/build.gradle.kts`
- [x] Update node packages: run the `updatepackages` script, e.g. by  `npm run updatepackages` to update versions in
  `package.json`
- [x] Run the `updatepackagelock`, e.g. by  `npm run updatepackagelock` script to update `package-lock.json` and check
  for security issues

### Dataland automated QA service

- [x] Update package versions in `dataland-automated-qa-service/requirements.txt`

### Dataland EuroDaT client

- [x] Check if the eurodatClientOpenApi.json in dataland-external-storage is in sync with the currently used version 
  of the client.
  Check on the https://eurodat.gitlab.io/trustee-platform/release_notes/ if there is a newer version available, if yes
  then update the version number used in docker-compose.
  Then start the eurodat client as described in the internal Dataland wiki and visit 
  http://localhost:8080/api/v1/client-controller/openapi 
  This should trigger an automated download of the openApiSpec. The content of this file will be in YAML format.
  Therefore convert its content to JSON with this converter: https://jsonformatter.org/yaml-to-json
  Then replace the content of eurodatClientOpenApi.json in the repo with the converted JSON from the converter.
  After doing this, there should be no diffs to main in eurodatClientOpenApi.json!
  If there are diffs, this means that it is out of sync and you should discuss this with someone on the MiNaBo team.

### Dataland Analytics

- [x] Update the backend API file
  - [x] Go to https://dataland.com/api/v3/api-docs/public and copy the backend api JSON
  - [x] Convert the JSON to a .yaml file preferably using an online-converter (https://jsonformatter.org/json-to-yaml)
  - [x] Paste the .yaml output into "datasets_open_api.yaml"
- [x] Accordingly, update the document API file
  - [x] Copy from https://dataland.com/documents/v3/api-docs/public to documents_open_api.yaml
- [x] Accordingly, update the community API file
  - [x] Copy from https://dataland.com/community/v3/api-docs/public to requests_open_api.yaml
- [x] Accordingly, update the QA API file
  - [x] Copy from https://dataland.com/qa/v3/api-docs/public to requests_open_api.yaml

### Dockerfile updates

Update versions in the following dockerfiles

- [x] `./dataland-api-key-manager/Dockerfile`
- [x] `./dataland-api-key-manager/DockerfileBase`
- [x] `./dataland-api-key-manager/DockerfileTest`
- [x] `./dataland-automated-qa-service/Dockerfile`
- [x] `./dataland-automated-qa-service/DockerfileBase`
- [x] `./dataland-automated-qa-service/DockerfileTest`
- [x] `./dataland-backend/Dockerfile`
- [x] `./dataland-backend/DockerfileBase`
- [x] `./dataland-backend/DockerfileTest`
- [x] `./dataland-batch-manager/Dockerfile`
- [x] `./dataland-batch-manager/DockerfileBase`
- [x] `./dataland-community-manager/Dockerfile`
- [x] `./dataland-community-manager/DockerfileBase`
- [x] `./dataland-community-manager/DockerfileTest`
- [x] `./dataland-document-manager/Dockerfile`
- [x] `./dataland-document-manager/DockerfileBase`
- [x] `./dataland-document-manager/DockerfileTest`
- [x] `./dataland-e2etests/Dockerfile`
- [x] `./dataland-e2etests/DockerfileBase`
- [x] `./dataland-email-service/Dockerfile`
- [x] `./dataland-email-service/DockerfileBase`
- [x] `./dataland-email-service/DockerfileTest`
- [x] `./dataland-frontend/Dockerfile`
- [x] `./dataland-frontend/DockerfileTest`
- [x] `./dataland-framework-toolbox/excel-to-csv/Dockerfile`
- [x] `./dataland-inbound-admin-proxy/Dockerfile`
- [x] `./dataland-inbound-proxy/Dockerfile`
- [x] `./dataland-internal-storage/Dockerfile`
- [x] `./dataland-internal-storage/DockerfileBase`
- [x] `./dataland-internal-storage/DockerfileTest`
- [x] `./dataland-keycloak/Dockerfile`  (also update realm json files with new version)
- [x] `./dataland-pgadmin/Dockerfile`
- [x] `./dataland-qa-service/Dockerfile`
- [x] `./dataland-qa-service/DockerfileBase`
- [x] `./dataland-qa-service/DockerfileTest`
- [x] `./dataland-rabbitmq/Dockerfile`
- [x] `./dataland-dummy-eurodat-client/DockerfileBase`
- [x] `./dataland-dummy-eurodat-client/DockerfileTest`
- [x] `./base-dockerfiles/DockerfileGradle`
- [x] Update the versions of the external images for api-key-manager-db, backend-db, keycloak-db, internal-storage-db,
  document-manager-db, qa-service-db, community-manager-db and frontend-dev in `./docker-compose.yml`
- [x] Check if there are any services in the `docker-compose.yml` file that have not gotten an update yet (e.g. a new
  service that is not covered by the tasks above)

## Server updates

Note: First manually create backups of the servers with a short retention duration.
Then start the update with one of the dev servers and deploy to it after the update.
If everything was fine, proceed with other servers. Double check for Prod.

Note: currently there seems to be an issue with the docker-compose plugin

Execute `sudo apt-get update && sudo apt-get upgrade` on

- [x] dev1.dataland.com
- [x] dev2.dataland.com
- [x] dev3.dataland.com
- [x] test.dataland.com
- [x] letsencrypt.dataland.com
- [x] (OPT) dataland.com

### ssh-keys maintenance

check that all ssh-keys are set and erased from people that have left

- [x] dev1.dataland.com
- [x] dev2.dataland.com
- [x] dev3.dataland.com
- [x] test.dataland.com
- [x] clone.dataland.com
- [x] letsencrypt.dataland.com
- [x] (OPT) dataland.com

## Check keycloak automatic logout if inactive

- [x] Check that you are automatically logged out if you are idle for too long and also get notified about this by a
  pop-up. Timeouts are defined as `TIME_BEFORE_REFRESH_TOKEN_EXPIRY_TO_DISPLAY_SESSION_WARNING_IN_MS`
  in `./dataland-frontend/src/utils/Constants.ts` and `ssoSessionIdleTimeout`
  in `./dataland-keycloak/realms/datalandsecurity-realm.json`.

## Check e-mails sent by keycloak

- [x] Check that the verification e-mail is sent and displayed correctly
- [x] Check that the reset password e-mail is sent and displayed correctly
- [x] Check that the password can be reset
- [x] Check that account linking via e-mail verification works correctly
- [x] Check that account linking via username & password verification works correctly

## Check e-mails sent by backend

- [x] Send an invitation (data) request from one of the dev servers and check if the e-mail to dataland@d-fine.com
  contains the right attachments and is displayed correctly.

## Check RabbitMQ dead letter queue and disk space

- [x] RabbitMQ does need at least 768MB of free disk space to operate. `ssh` into all servers and check the available
  disk space with `df -h` command. If the open disk space is close to the minimum requirement, clear up disk space
  with `sudo docker image prune --all`.
- [x] On all environments, no new messages should have been added to the dead letter queue since the last manual
  maintenance. If new messages have appeared this needs to be investigated. The dead letter queue can be accessed
  and messages on it read in the RabbitMQ GUI. Access it by port-forwarding port `6789` from the server and then
  accessing the GUI at `localhost:6789/rabbitmq`. After login, the dead letter queue can be found at Queues &rarr;
  deadLetterQueue &rarr; Get message.

## Check that links to external webpages are working
- [x] Go to the frameworks/upload page and check that the link to WWF Pathway to Paris is still working as intended.

## Check that the main branch has no sonar issues
- [x] Go to the sonar report summary of the main branch and verify that there are no sonar findings. If there are sonar 
  findings fix them.

## Conclusion

- [x] After updating all components check if everything still works
- [x] The new version is deployed to a dev server using this branch
    - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
    - [x] It's verified that everything seems to be working fine by manually using the website
    - [x] All implemented Social Logins have been tested manually in the UI
- [x] This template and the internal wiki page have been updated to reflect the latest state of tasks required and known issues with upgrades
- [ ] The Merge Request commit message needs to contain 'manual maintenance' to satisfy the CI maintenance check in
  future commits

# Review (to be completed by the reviewer)

- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github.
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described above
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] It's verified that everything seems to be working fine by manually using the website
  - [x] All implemented Social Logins have been tested manually in the UI
  - [x] The "mailto:*" buttons ("I am interested" and "Get in touch") on the landing page work.
  - [x] Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
    authorization has worked by assuring that you get a 200 response
- [x] If any work on the UI is to be merged, those changes were also documented in the Figma
- [x] The local Dev stack still works: execute `startDevelopmentStack.sh` and `npm run testcomponent` in dataland-frontend (a bunch of cypress frontend e2e tests fails locally without manually clicking away the cookie banner => meaningless to run testpipeline)
- [x] After(!) the cypress tests have passed locally, execute the backend-e2e-tests `./gradlew dataland-e2etests:test`
- [x] Locally: Go to the swagger-UI, authorize, run a "GET" request to the companies endpoint and assure that your
  authorization has worked by assuring that you get a 200 response
- [x] It is assured that deploying this feature branch over the current main does not break anything
  - [x] A fresh clone of dataland.com is generated (see Wiki page on "OTC" for details)
  - [x] Deploy the feature branch to clone with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [x] Verify that the CD run is green
  - [x] Verify that everything seems to be working fine by manually using the website
- [x] Merge using Squash Commit. The Merge Commit Message needs to contain "Manual Maintenance"
- [ ] After merge check SonarQube state of main branch at https://sonarcloud.io/summary/new_code?id=d-fine_Dataland. 
  The full scan might reveal new issues (e.g. deprecation) on old code which is generally not detected on the branch.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 1]
```diff
@@ -15,7 +15,6 @@ If an issue arises from a new update that cannot be solved in the scope of MM, a
 
 ### Gradle update
 
-- [ ] Execute `gradlew dependencyUpdates` to get a report on Dependencies with updates
```
Comment (MarcTM01) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:18]: Brauchen wir das nicht mehr, weil wir davon ausgehen, dass `refreshVersions` das alles macht? Oder wieso kann das weg?

File: dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts [hunk 2]
```diff
@@ -80,7 +80,7 @@ describe(
             },
             () => {
               const expectedNumberOfCompanies = fixtureData.length;
-              cy.wait(2000);
+              cy.wait(5000);
```
Comment (MarcTM01) [inline dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts:83]: Ich denke, dass ist dafür da, um die flakyness von dem 49/50 companies test zu reduzieren, oder?

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -48,31 +48,35 @@ junit-jupiter = "5.10.2"
 
 gradle-tooling = "org.gradle:gradle-tooling-api:8.7"
 
-detekt-cli = "io.gitlab.arturbosch.detekt:detekt-cli:_"
```
Comment (MarcTM01) [inline gradle/libs.versions.toml:51]: Warum werden hier überall die `_` durch konkrete versionen ersetzt?

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -196,3 +204,48 @@ poi-ooxml = "org.apache.poi:poi-ooxml:5.2.5"
...
+
+spring-data-jpa = "org.springframework.data:spring-data-jpa:3.2.5"
+
+spring-beans = "org.springframework:spring-beans:6.1.6"
+
+spring-security-config = "org.springframework.security:spring-security-config:6.2.4"
+
+spring-boot-autoconfigure = "org.springframework.boot:spring-boot-autoconfigure:3.2.5"
+
+spring-web = "org.springframework:spring-web:6.1.6"
```
Comment (MarcTM01) [inline gradle/libs.versions.toml:239]: Ich dachte, dass gralde-spring plugin kümmert sich um die Versionen von all dem Spring-Zeugs.

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 1 above)
Comment (jhaerDfine) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:18]: gute frage, wurde von Siegfried geändert:
https://github.com/d-fine/Dataland/pull/479/commits/edec7683a746425a7eb490867b12084ea22d667d

File: dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts (hunk 2 above)
Comment (jhaerDfine) [reply dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts:83]: das ist auch von Siegfried, aber ja das war die idee um die flakyness zu reduzieren

File: gradle/libs.versions.toml (hunk 3 above)
Comment (jhaerDfine) [reply gradle/libs.versions.toml:51]: ebenfalls von Siegfried

File: gradle/libs.versions.toml (hunk 4 above)
Comment (jhaerDfine) [reply gradle/libs.versions.toml:239]: ebenfalls von Siegfried

File: gradle/libs.versions.toml (hunk 3 above)
Comment (MarcTM01) [reply gradle/libs.versions.toml]: Besprochen und abgeändert

File: gradle/libs.versions.toml (hunk 4 above)
Comment (MarcTM01) [reply gradle/libs.versions.toml]: Besprochen und abgeändert

**Pull request merged**